domains/client/relayer/src/worker.rs

```diff
@@ -174,50 +219,32 @@ where
         + AuxStore
         + ProofProvider<Block>
         + ProvideRuntimeApi<Block>,
-    Client::Api: RelayerApi<Block, NumberFor<Block>>,
-    MP: Fn(&Arc<Client>, Block::Hash) -> Result<(), Error>,
+    MP: Fn(&Arc<Client>, (NumberFor<Block>, Block::Hash), ExtraData) -> Result<(), Error>,
     SO: SyncOracle,
-    CRM: Fn(ChainId, NumberFor<Block>) -> Result<bool, ApiError>,
+    CRM: Fn(NumberFor<Block>) -> Result<Option<ExtraData>, ApiError>,
 {
-    let chain_id = Relayer::chain_id(&client)?;
     tracing::info!(
         target: LOG_TARGET,
         "Starting relayer for chain: {:?}",
         chain_id,
     );
-    let mut chain_block_import = client.import_notification_stream();
+    let mut chain_block_finalization = client.finality_notification_stream();
 
     // from the start block, start processing all the messages assigned
-    // wait for new block import of chain,
+    // wait for new block finalization of the chain,
     // then fetch new messages in the block
     // construct proof of each message to be relayed
     // submit XDM as unsigned extrinsic.
-    while let Some(block) = chain_block_import.next().await {
+    while let Some(block) = chain_block_finalization.next().await {
```

Comment:
IIUC there only be one consensus block finalization notification per segment, so the finalized consensus block will not come like `#1,#2,..` but instead `#1000,#2000,..` cc @nazar-pc, perhaps use the block import notification with something like `block_to_process = imported_block - K`

Reply:
That is fine. Relayer does not expect the block come iincremental. We just need the latest finalized block and we process all the blocks that in between last processed to latest finalized.

Reply:
This may cause problem, if there is one new segment per 1 hour then all relayers from any domain will idle for 1 hour then start relaying XDM at the same time, which brings additional delay and network congestion.

Reply:
Yes that is correct, XDM is already quiet delayed due to block confirmation and we expliclty want to wait until consensus block is finalized.

> notification with something like block_to_process = imported_block - K

This approach defeats the purpose of waiting for consensus block finalization but rather relying on Kdepth. I'm not sure why we moved away from it but I would like to still relay on finalization instead of block_import

Reply:
I agree using block finalization is better in term of readability, but the consensus block finalization is coupled with archiving and differs from what is provided by substrate by default. Besides the above concern about additional delay and network congestion, also consider if an XDM is dropped silently it needs to wait for another new segment before the next retry.

Reply:
> also consider if an XDM is dropped silently it needs to wait for another new segment before the next retry.

I dont see any concern here. There will always be a retry of the XDM in the next turn and XDM will end being included in the chain. This delay is what we opted for as a base XDM using confirmation and finalization. 
We expect external services to be built on top to speed up these transfers at higher incentive and is not a concern for the Subspace.

BUt feel free to propose any potential alternatives on making this approach faster and I would like to understand what can be done here better

Reply:
>There will always be a retry of the XDM in the next turn and XDM will end being included in the chain.

The difference is one retry per block versus one retry per segment.

>This delay is what we opted for as a base XDM using confirmation and finalization.
>BUt feel free to propose any potential alternatives on making this approach faster and I would like to understand what can be done here better

IMO, the hard required delay is: domain block confirmation (challenge period) + consensus block confirmation (K block depth) while the consensus block finalization is: K blocks depth + blocks of one segment. Simply using the consensus block import notification + K block depth can make it faster, non-blocker.

Reply:
From my understanding, K blocks depth is when a consensus block is confirmed and inrevertable which is what we need for XDM (and also used in many other places), while finalization is coupled with the archiving process to ensure the block is not pruned before archiving, finalization take much longer than K blocks but doesn't mean confirmed and non-finalized block is revertable. cc @nazar-pc in case my understanding is wrong.

Non-blocker for using finalization block as the additional delay is not a concern, plz resolve the conflict though.